# Fuzzing: Breaking Things with Random Inputs
>原文：https://www.fuzzingbook.org/html/Fuzzer.html

在本节中，我们将从最简单的测试生成技术开始。随机文本生成（也称fuzzing）的主要思想是向程序中输入一串随机字符，希望从中发现故障。



## A Testing Assignment

模糊测试诞生于“1988年秋天一个黑暗又风雨交加的夜晚”[Takanena et al,2008](http://us.artechhouse.com/Fuzzing-for-Software-Security-Testing-and-Quality-Assurance-Second-Edition-P1930.aspx)。 威斯康星州麦迪逊市的巴顿·米勒教授坐在他的公寓里,通过1200波特的电话线与他的计算机相连接。 暴风雨在线路上造成了噪音,而这种噪音反过来又导致了UNIX命令在计算机上受到异常输入并崩溃。 频繁的崩溃让他感到惊讶——程序不应该更健壮一些吗? 作为一名科学家,他想研究这个问题的严重程度及其原因。 因此,他为威斯康星大学麦迪逊分校的学生设计了一个编程练习——一个让学生创建第一批模糊测试工具的练习。

 [这个项目](http://pages.cs.wisc.edu/~bart/fuzz/CS736-Projects-f1988.pdf) 中提到：

> 本项目的目标是在输入流不可预测的情况下，评估各种 UNIX 实用程序的鲁棒性。[...]首先，您将构建一个模糊生成器。这是一个会输出随机字符流的程序。其次，您将使用模糊发生器攻击尽可能多的 UNIX 实用程序，目的是破解它们。

这项任务抓住了模糊测试的本质：创建随机输入，看看它们是否会破坏一些东西。只要让它运行足够长的时间，你就会发现。

## A Simple Fuzzer

尝试完成上面这个简单的项目，构建一个模糊测试生成器。思路是生成随机字符,将它们添加到一个缓冲字符串变量(`out`)中,最后返回这个字符串。

这个实现使用了以下Python特性和函数:

- `random.randrange(a, b)` –返回a和b之间的随机数 
- `range(a, b)` – 创建一个区间在a和b之间的整数列表(可以被当作迭代器使用)
- `for elem in list: body` – 以循环的方式执行`body`，`elem`依次取`list`中的每个值
- `for i in range(start, end): body` – 以循环的方式执行`body`,`i`从`start`取到`end` 
- `chr(n)` – 返回ASCII码`n`的字符

```c
import random
```

导入random模块后，下面是实际`fuzzer`函数

```c
def fuzzer(max_length: int = 100, char_start: int = 32, char_range: int = 32) -> str:
    string_length = random.randrange(0, max_length + 1)
    out = ""
    for i in range(0, string_length):
        out += chr(random.randrange(char_start, char_start + char_range))
    return out
```

- 指定生成字符串的最大长度：`max_length: int = 100`
- 指定起始ASCII字符：`char_start: int = 32` 
- 指定取值范围：`char_range: int = 32`，从`char_start`开始,默认值为32

在默认参数下，`fuzzer()`函数返回一串随机字符：

```c
'!7#%"*#0=)$;%6*;>638:*>80"=</>(/*:-(2<4 !:5*6856&?""11<7+%<%7,4.8,*+&,,$,."'
```

Bart Miller）创造了”fuzz“一词，来指代这种随机的、无结构的数据，现在想象一下,如果这个“fuzz”字符串作为输入，被一个期待特定输入格式的程序使用。比如,一个逗号分隔的值列表,或者一个电子邮件地址。这个程序能够无问题地处理这样的输入吗?

我们可以很容易的进行设置fuzzing，来产生其他类型的输入，例如，生成一长串小写字母，使用`ord("a")`来返回字母a的ASCII码，返回的值作为fuzzer函数的第二个参数

```c
fuzzer(1000, ord('a'), 26)
```

```c
'zskscocrxllosagkvaszlngpysurezehvcqcghygphnhonehczraznkibltfmocxddoxcmrvatcleysksodzlwmzdndoxrjfqigjhqjxkblyrtoaydlwwisrvxtxsejhfbnforvlfisojqaktcxpmjqsfsycisoexjctydzxzzutukdztxvdpqbjuqmsectwjvylvbixzfmqiabdnihqagsvlyxwxxconminadcaqjdzcnzfjlwccyudmdfceiepwvyggepjxoeqaqbjzvmjdlebxqvehkmlevoofjlilegieeihmetjappbisqgrjhglzgffqrdqcwfmmwqecxlqfpvgtvcddvmwkplmwadgiyckrfjddxnegvmxravaunzwhpfpyzuyyavwwtgykwfszasvlbwojetvcygectelwkputfczgsfsbclnkzzcjfywitooygjwqujseflqyvqgyzpvknddzemkegrjjrshbouqxcmixnqhgsgdwgzwzmgzfajymbcfezqxndbmzwnxjeevgtpjtcwgbzptozflrwvuopohbvpmpaifnyyfvbzzdsdlznusarkmmtazptbjbqdkrsnrpgdffemnpehoapiiudokczwrvpsonybfpaeyorrgjdmgvkvupdtkrequicexqkoikygepawmwsdcrhivoegynnhodfhryeqbebtbqnwhogdfrsrksntqjbocvislhgrgchkhpaiugpbdygwkhrtyniufabdnqhtnwreiascfvmuhettfpbowbjadfxnbtzhobnxsnf'
```

## Fuzzing External Programs

将生成的数据作为参数给其他的程序，会发生什么呢？

## Creating Input Files

创建一个临时文件 

```c
import os
import tempfile

basename = "input.txt"
tempdir = tempfile.mkdtemp()
FILE = os.path.join(tempdir, basename)
print(FILE)
```

```c
C:\Users\Kksa\AppData\Local\Temp\tmphb3t_72z\input.txt
```

现在可以打开这个文件进行写入。Python的`open()`函数可以打开一个文件,这样我们就可以向文件中写入任意内容。它通常和`with`语句一起使用,可以确保一旦文件不再需要就关闭文件。

```c
data = fuzzer()
with open(FILE, "w") as f:
    f.write(data)
```

通过读取文件内容来确定是否写入：

```c
contents = open(FILE).read()
print(contents)
assert(contents == data)
```

```c
C:\Users\Kksa\AppData\Local\Temp\tmpm9b92_zu\input.txt
?&0+8& 1!;%< .%>?"<=:,5<-1+;?)).%)3# ,+9=<+ 3'+2 ,'.0!!269$3!>
```

## Invoking(调用) External Programs

已经有了一个输入文件，测试外部程序，这里测试一下`bc`计算器程序，为了调用 `bc,`我们可以使用 Python 的 `subprocess` 模块。它的用法如下:

> 由于这里调用的是Linux下的bc计算器，所以接下来用linux进行操作，有的Linux未安装bc，需要自己安装

```c
program = "bc"
with open(FILE, "w") as f:
    f.write("2 + 2\n")
result = subprocess.run([program, FILE],
                        stdin=subprocess.DEVNULL,
                        stdout=subprocess.PIPE,
                        stderr=subprocess.PIPE,
                        universal_newlines=True)  
```

这里使用的是`Thonny`进行Linux下的python代码编写。

![image](https://github.com/kksaert/fuzzbook/assets/83335903/cb9ea5fc-d2c8-421b-bff4-02b62cf3b6f8)


还可以检查状态，值为0表示程序正确终止

```c
result.returncode
0
```

任何错误信息都将显示在 `results.stderr` 中：

实际上，可以输入任何您喜欢的程序来代替 `bc`。不过要注意的是，如果你的程序能够改变甚至破坏你的系统，那么模糊输入中包含的数据或命令就有很大的风险。

假设您要测试一个文件删除程序，例如 `rm -fr FILE`，其中FILE是由fuzzer()生成的字符串。fuzzer()（使用默认参数）产生的 FILE 参数导致删除所有文件的几率有多大？

这种几率实际上比你想象的要高。例如，如果删除 `/`（所有文件的根目录），你的整个文件系统就会消失。如果删除 `~`（你的主目录），你的所有文件都将不复存在。如果删除 `.（`当前文件夹），当前目录下的所有文件都将消失。要产生上述任何一个结果，都需要字符串长度为 1（几率：100 分之 1）和这三个字符中的一个（几率：32 分之 3），这确实是千分之一的几率。

```c
1/100 * 3/32
0.0009375
```

不过，实际上只要第二个字符是空格，我们就可以处理任何字符串，毕竟 `rm -fr / Whatever` 会先处理 `/`，然后才处理后面的字符。第一个字符的几率是 32 分之 3，空格的几率是 32 分之 1，所以我们的几率是 300 分之 1：

```c
3/32 * 1/32
0.0029296875
```

鉴于模糊测试通常会运行数百万次，您确实不想冒这个风险。在可以随意重置的安全环境中运行模糊器，例如 Docker容器。

## Long-Running Fuzzing

现在让我们将大量输入，输入到我们测试的程序中，看看它是否会在某些情况下崩溃。我们将所有结果作为输入数据和实际结果对存储在运行变量中。 （注意：运行这可能需要一段时间。）

```c
trials = 100
program = "bc"

runs = []

for i in range(trials):
    data = fuzzer()
    with open(FILE, "w") as f:
        f.write(data)
    result = subprocess.run([program, FILE],
                            stdin=subprocess.DEVNULL,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE,
                            universal_newlines=True)
    runs.append((data, result))
```

我们现在可以查询`runs`来获取一些统计信息。例如,我们可以查询通过的运行(即没有错误信息)有多少次。这里我们使用了一个列表推导式:形式为 expression``for` element `in` list `if` condition。 会返回一个由每个满足`condition` 条件的`element` 元素经过`expression` 运算后的列表。(实际上,列表推导式会返回一个列表生成器,但是对我们目的来说,生成器表现得像一个列表。) 在这里,对于所有满足条件的元素,我们让`expression` 为1,然后用 `sum()` 对列表中所有元素求和。

简单来说,就是通过列表推导式筛选出所有没有错误的运行,对其指定表达式为1,然后用sum()求和,即可以统计通过的运行次数。

> 这里就是把每个数据以及对应的运行结果存储到一个二维数组值中，然后用推导式找出没问题的

> 关于Python中的推导式参考:[讲解一下Python推导式](https://www.lxlinux.net/8507.html#:~:text=列表推导式是一种快速生成列表的方式。 它一般用“ []”括起来，例如 >>> [i for i in,4%2C 9%2C 16%2C 25%2C 36%2C 49%2C 64%2C 81])

![image](https://github.com/kksaert/fuzzbook/assets/83335903/a95a0e61-6c92-4446-901c-9efeca17513c)


由于是fuzz，每次的统计结果可能不一样

大多数输入显然是无效的--这并不奇怪，因为随机输入不太可能包含有效的算术表达式。

让我们看看第一条错误信息：

```c
errors = [(data, result) for (data, result) in runs if result.stderr != ""]
(first_data, first_result) = errors[0]

print(repr(first_data))
print(first_result.stderr)
```

![image](https://github.com/kksaert/fuzzbook/assets/83335903/0f40cd5b-c329-455c-9a6b-e47f50002b4f)


 除了非法字符、解析错误或语法错误之外，还有其他运行信息吗？(比如，崩溃或发现致命错误之类的？

```c
[result.stderr for (data, result) in runs if 
result.stderr != ""
and "illegal character" not in result.stderr
and "parse error" not in result.stderr 
and "syntax error" not in result.stderr]
```

也许bc就表示崩溃了。但是，返回代码从来都不是非零：

```c
sum(1 for (data, result) in runs if result.returncode != 0)
91
```

> 这里我实际测试的时候代码是跑不动的，但是这不重要，学习思路就行



## Bugs Fuzzers Find

1989年，当米勒和他的学生运行他们的第一批模糊器时，他们发现了一个令人震惊的结果：他们模糊过的UNIX实用程序中，约有三分之一出现了问题--在面对模糊输入时，它们会崩溃、挂起或以其他方式失效，这其中也包括上文提到的bc程序。(显然，这些漏洞现在已经被修复了！）。

考虑到这些UNIX实用程序中的许多都用于处理网络输入的脚本中，这一结果令人震惊。程序员们迅速构建并运行了自己的模糊器，匆忙修复了报告的错误，并学会不再相信外部输入。

米勒的模糊实验发现了哪些问题？事实证明，程序员在1990年所犯的错误与今天所犯的错误如出一辙。

## Buffer Overflows

许多程序都内置了输入和输入元素的最大长度。在C语言中，很容易在程序（或程序员）毫无察觉的情况下超过这些长度，从而引发所谓的`缓冲区溢出`。例如，即使输入字符串超过 8 个字符，下面的代码也能轻松地将输入字符串复制到工作日字符串中：

```c
char weekday[9]; // 8 characters + trailing '\0' terminator
strcpy (weekday, input);
```

具有讽刺意味的是,如果输入为“Wednesday”(9个字符)就已经会失败了;任何额外的字符(这里是‘y’和随后的‘\\0’字符串终止符)会被简单地复制到weekday之后内存中的任意内容,触发任意行为;可能是某个布尔字符变量,它会从‘n’被设置为‘y’。通过模糊测试,产生任意长度的输入和输入元素非常容易。

总体来说,由于weekday数组长度不足,导致向后内存写入多余输入时可能会影响其他变量的值,进而导致程序异常行为。模糊测试可以轻松产生超长或异常输入,因此更容易发现这类问题。这启示我们在编程时需要注意数组边界检查,对输入长度进行限制,避免向后内存越界写入影响其他数据。

我们可以轻松地在 Python 函数中模拟这种缓冲区溢出行为：

```c
def crash_if_too_long(s):
    buffer = "Thursday"
    if len(s) > len(buffer):
        raise ValueError
```

继续用上面的fuzzer生成的数据当作输入，程序很快就崩溃了

```c
from ExpectError import ExpectError
    
trials = 100
with ExpectError():
    for i in range(trials):
        s = fuzzer()
        crash_if_too_long(s)
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/292568387.py", line 5, in <cell line: 2>
    crash_if_too_long(s)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/2784561514.py", line 4, in crash_if_too_long
    raise ValueError
ValueError (expected)
```

上述代码中的 `ExpectError()` 确保了错误信息被打印出来，但执行仍在继续；这是为了将这个 "预期 "错误与其他代码示例中的 "意外 "错误区分开来。

## Missing Error Checks

许多编程语言没有异常处理机制，而是让函数在特殊情况下返回特殊错误代码。例如 C 函数getchar()，从标准输入 stdin 获取一个字符（一个无符号字符）；如果到达文件末尾或发生读错误，则返回 EOF。现在假设程序员正在扫描输入以查找下一个字符，使用getchar()读入字符，直到读取到空格字符：

```c
while (getchar() != ' ');
```

猜想如果输入在预想之前结束（模糊测试完全可能生成这种结果），会发生什么情况？好吧，`getchar()`会返回`EOF`，并在再次调用时继续返回`EOF`；因此，上面的代码将会进入一个无限循环。

同样的,我们可以模拟这种缺少错误检查的情况。下面是一个函数,如果输入中没有空格,它将会有效地挂起:

```c
def hang_if_no_space(s):
    i = 0
    while True:
        if i < len(s):
            if s[i] == ' ':
                break
        i += 1
```

利用之前学习的超时机制，可以在函数运行一段时间后中断按，

```c
from ExpectError import ExpectTimeout

trials = 100
with ExpectTimeout(2):
    for i in range(trials):
        s = fuzzer()
        hang_if_no_space(s)
```

报错如下：

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/3194687366.py", line 5, in <cell line: 2>
    hang_if_no_space(s)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/3035466707.py", line 3, in hang_if_no_space
    while True:
  File "/Users/zeller/Projects/fuzzingbook/notebooks/Timeout.ipynb", line 43, in timeout_handler
    raise TimeoutError()
TimeoutError (expected)
```



## Rogue Numbers

通过fuzzing，很容易在输入中生成不常见的值，从而导致各种有趣的行为。请看下面这段同样使用 C 语言编写的代码，它首先从输入中读取一个缓冲区的大小，然后分配一个给定大小的缓冲区：

```c
char *read_input() {
    size_t size = read_buffer_size();
    char *buffer = (char *)malloc(size);
    return (buffer);
}
```

> - 调用read_buffer_size()获取需要读入的数据大小size，
> - 使用malloc()函数动态分配size个字节大小的内存空间,用于存储输入数据
> - 将分配的内存空间首地址强制转换为`char*` 类型的指针,存储在buffer变量中
> - 函数返回一个buffer指针，该指针指向存储输入数据的缓冲区

同样，我们也可以在 Python 中轻松地模拟这种效果。如果传递的值（字符串）在转换为整数后过大，函数 `collapse_if_too_large()` 就会失败。

```c
def collapse_if_too_large(s):
    if int(s) > 1000:
        raise ValueError
```

可以`fuzzer()`函数创造一个数字字符串

```c
long_number = fuzzer(100, ord('0'), 10)
print(long_number)
7056414967099541967374507745748918952640135045
```

如果向上面的`collapse_if_too_large(s)`函数输入fuzzer生成的数字符串，很快就会报错

```c
with ExpectError():
    collapse_if_too_large(long_number)
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/2775103647.py", line 2, in <cell line: 1>
    collapse_if_too_large(long_number)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/1591744602.py", line 3, in collapse_if_too_large
    raise ValueError
ValueError (expected)
```



## Catching Errors

当米勒和他的学生构建他们的第一个fuzzer时，他们可以识别错误，因为程序会崩溃或挂起 - 这两个条件很容易识别。但是，如果故障更微妙，我们需要更深入的检查。

## Generic Checkers

如上所述，缓冲区溢出是一个非常的普遍问题：在 C 和 C++ 等语言中，程序可以访问其内存的任意部分——甚至是那些未初始化、已经释放或根本属于数据部分的部分。如果你想编写一个操作系统，这是必要的，如果你想要最大的性能或控制，这是很好的，但如果你想避免错误，那就太糟糕了。幸运的是，有一些工具可以帮助在运行时捕获此类问题，并且与模糊测试结合使用时非常有用。

## Checking Memory Accesses

为了在测试期间捕获有问题的内存访问，可以在特殊的内存检查环境中运行C程序。在运行时，它们检查每个内存操作是否访问有效和初始化的内存。一个流行的例子是[LLVM Address Sanitizer](https://clang.llvm.org/docs/AddressSanitizer.html)，它检测整套潜在的危险内存安全违规行为。在下面的示例中，我们将使用该工具编译一个相当简单的C程序，并通过读取内存的已分配部分来引发越界读取。

```c
#include <stdlib.h>
#include <string.h>

int main(int argc, char** argv) {  
    char *buf = malloc(100); //分配100个字节的内存空间,赋值给指针buf
    memset(buf, 42, 100); //将buf所指内存空间的前100个字节都设置为值42

    int index = atoi(argv[1]); //从命令行参数argv[1]读取一个索引index
    char val = buf[index]; //访问buf中的第index个元素,保存到val

    free(buf); //使用free()释放内存
    return val; //返回val的值
}
```

> argc表示传入参数的个数(参数计数)
> argv数组存储这些参数,argv[0]通常是程序名
> 剩余argv[1]、argv[2]...存储了命令行传入的参数
>
> 那么在上面的main函数中:
>
> - argc 是2
> - argv[0] 是 "./program"
> - argv[1] 是 "99"

我们使用地址sanitization(地址错误检测)启用的方式编译这个C程序后并运行：

```c
clang -fsanitize=address -g -o program program.c

./program 99; echo $?

42
```

> 在Linux/Unix系统中,每一个进程结束时都会返回一个退出状态码(exit status code),表示程序执行完成后的状态。
>
> 这个退出码是一个整数值,缺省为0表示正常退出。非0的值通常表示不同的错误条件。
>
> 该退出码可以通过`$?`变量获取上一个运行进程的返回值。

> LLVM Address Sanitizer 这个工具，clang和gcc都自带了，直接用  

![image](https://github.com/kksaert/fuzzbook/assets/83335903/bf357cd1-87d1-43dd-8f18-3996de09fcce)


但是，访问buf[110]会导致AddressSanitizer出现越界错误。

```c
==93901==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60b0000000ae at pc 0x5602b0a6af4c bp 0x7fff7f3f08f0 sp 0x7fff7f3f08e8
READ of size 1 at 0x60b0000000ae thread T0
    #0 0x5602b0a6af4b in main /root/桌面/program.c:11:16
    #1 0x7fed1f5a06c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #2 0x7fed1f5a0784 in __libc_start_main csu/../csu/libc-start.c:360:3
    #3 0x5602b09ad300 in _start (/root/桌面/program+0x20300) (BuildId: 06bffe530a0e1fa2c94df45661590401316da976)

0x60b0000000ae is located 10 bytes to the right of 100-byte region [0x60b000000040,0x60b0000000a4)
allocated by thread T0 here:
    #0 0x5602b0a3014e in __interceptor_malloc (/root/桌面/program+0xa314e) (BuildId: 06bffe530a0e1fa2c94df45661590401316da976)
    #1 0x5602b0a6aebf in main /root/桌面/program.c:6:17
    #2 0x7fed1f5a06c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16

SUMMARY: AddressSanitizer: heap-buffer-overflow /root/桌面/program.c:11:16 in main
Shadow bytes around the buggy address:
  0x0c167fff7fc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff8000: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
=>0x0c167fff8010: 00 00 00 00 04[fa]fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8020: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8030: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==93901==ABORTING

```

如果要查找c程序中的漏洞，进行fuzzing是非常简单的，虽然会降低被检测程序运行速度(因工具而异，AddressSanitizer通常是2倍)，并且消耗更多内存,但是与人工花费在找到这些bug上的努力相比,CPU可以说是非常便宜的。

对内存的越界访问是一个巨大的安全风险,因为它们可能让攻击者访问或甚至修改那些不应该让他们访问的信息。一个著名的例子是[HeartBleed bug](https://en.wikipedia.org/wiki/Heartbleed),这是一个OpenSSL库中的一个安全漏洞,OpenSSL库实现了在计算机网络上提供通信安全的加密协议。

HeartBleed漏洞是通过向SSL心跳服务发送一个特制的命令来利用的。心跳服务被用来检查服务器那一端是否还活着。一个客户端会发送一个字符串给服务,比如:

```c
BIRD(4个字母)
```

服务器然后会回复'BIRD',这样客户端就知道服务器还活着。

在 OpenSSL 的实现中，这些内存内容可能涉及加密证书、私钥等，更糟糕的是，没有人会注意到这些内存刚刚被访问过。当HeartBleed漏洞被发现时，它已经存在多年，没有人会知道是否以及哪些秘密已经泄露；[HeartBleed](http://heartbleed.com/)的公告页面说明了一切。

但是，HeartBleed 是如何被发现的呢？非常简单。Codenomicon 公司和谷歌公司的研究人员用`memory sanitizer`编译了OpenSSL库，然后向其中灌输模糊测试命令。然后，`memory sanitize`就会注意到是否发生了越界内存访问--实际上，它很快就会发现这一点。

内存检查器只是模糊测试过程中检测运行时错误的众多检查器之一。在 [chapter on mining function specifications](https://www.fuzzingbook.org/html/DynamicInvariants.html)中，我们将进一步了解如何定义通用检查器。



## Information Leaks

信息泄漏不仅可能通过非法内存访问发生，也可能发生在"有效 "内存中--如果该 "有效 "内存包含不应泄漏的敏感信息。让我们用 Python 程序来说明这个问题。首先，让我们创建一些充满实际数据和随机数据的程序内存：

```C
secrets = ("<space for reply>" + fuzzer(100) +
           "<secret-certificate>" + fuzzer(100) +
           "<secret-key>" + fuzzer(100) + "<other-secrets>")
```

我们在`secrets`中添加更多的“内存”字符，用"deadbeef"作为未初始化内存的标记：

```c
uninitialized_memory_marker = "deadbeef"
while len(secrets) < 2048:
    secrets += uninitialized_memory_marker
```

接下来定义了一个服务(类似于上面的心跳服务)，该服务将接收答复并发送回一个长度。它会将要发送的回复存储在内存中，然后以给定的长度发送回去。

```c
def heartbeat(reply: str, length: int, memory: str) -> str:
    # Store reply in memory
    memory = reply + memory[len(reply):]

    # Send back heartbeat
    s = ""
    for i in range(length):
        s += memory[i]
    return s
```

> - 首先将`reply`的内容附加到`memory`之前的部分，这个操作使用了字符串切片，意味着从 `memory` 中丢弃了之前的部分，只保留了与 `reply` 长度相同的部分。
> - 通过一个循环来构建一个新的字符串 `s`。在循环中，从 `memory` 的开头开始，逐个字符地将其添加到 `s` 中，直到达到指定的 `length`
> - 最终，函数返回了构建好的字符串 `s`，这个字符串是从存储中提取的内容，长度不超过指定的 `length`。

这在处理标准字符串时完全正常：

```c
heartbeat("potato", 6, memory=secrets)
'potato'

heartbeat("bird", 4, memory=secrets)
'bird'
```

然而，如果长度大于回复字符串的长度，内存的额外内容会泄漏出来。需要注意的是，所有这些仍然都在正常的数组边界内，因此不会触发`address sanitizer`：

```c
heartbeat("hat", 500, memory=secrets)

'hatace for reply>#,,!3?30>#61)$4--8=<7)4 )03/%,5+! "4)0?.9+?3();<42?=?0<secret-certificate>7(+/+((1)#/0\'4!>/<#=78%6$!!$<-"3"\'-?1?85!05629%/); *)1\'/=9%<secret-key>.(#.4%<other-secrets>deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadb'
```

如何检测这些问题？核心思想是识别不应该泄漏的信息，比如给定的秘密，还有未初始化的内存。我们可以在一个小的Python示例中模拟这种检查：

```c
from ExpectError import ExpectError
with ExpectError():
    for i in range(10):
        s = heartbeat(fuzzer(), random.randint(1, 500), memory=secrets)
        assert not s.find(uninitialized_memory_marker)
        assert not s.find("secret")
```

报错如下：

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/4040656327.py", line 4, in <cell line: 1>
    assert not s.find(uninitialized_memory_marker)
AssertionError (expected)
```

通过这样的检查，我们可以发现secrets或未初始化的内存会泄露出来， 在[chapter on information flow](https://www.fuzzingbook.org/html/InformationFlow.html)中，我们将会介绍如何自动执行此操作，将敏感信息和从他们派生出来的值标记为“受污染”，并确保“受污染”值不会外泄。

作为一个经验之谈，在fuzzing期间，应该尽可能多的进行自动化检查，CPU是廉价的，而错误是昂贵的，如果仅执行程序而不检查错误，你将会错失很多机会。



## Program-Specific Checkers

除了适用于特定平台(或语言)程序的通用检查器外，您还可以设计适用于您的程序或子系统的特定检查器。在[Introduction to Software Testing](https://www.fuzzingbook.org/html/Intro_Testing.html)这一章中，我们已经提到了运行时验证方面的技术，可以在运行时检查函数的结果是否正确。

一种早期检测错误的关键思想是进行断言 -检查重要函数的输入（前置条件）和结果（后置条件）的断言。您在程序中使用的断言越多，就越有可能在执行过程中检测到由通用检查器无法检测到的错误 - 尤其是在模糊测试期间。如果您担心断言对性能的影响，要记住在生产环境中可以关闭断言。

对于查找错误，断言的最重要的用途之一是检查复杂数据结构的完整性。让我们通过一个简单的例子来说明这个概念。假设我们有一个机场代码到机场的映射，就像下面这样：

```c
airport_codes: Dict[str, str] = {
    "YVR": "Vancouver",
    "JFK": "New York-JFK",
    "CDG": "Paris-Charles de Gaulle",
    "CAI": "Cairo",
    "LED": "St. Petersburg",
    "PEK": "Beijing",
    "HND": "Tokyo-Haneda",
    "AKL": "Auckland"
}  # plus many more
```

```c
airport_codes["YVR"]
'Vancouver'

"AKL" in airport_codes
True
```

这个机场代码列表非常关键：如果在任何机场代码中存在拼写错误，这可能会影响我们所拥有的任何应用程序。因此，我们引入一个函数来检查列表的一致性。这种一致性条件称为表示不变性，检查它的函数（或方法）通常被命名为 `repOK()`，表示“表示是正确的”。

首先，我们来编写一个用于单个机场代码的检查器。如果代码不一致，检查器将失败。

```c
def code_repOK(code: str) -> bool:
    assert len(code) == 3, "Airport code must have three characters: " + repr(code)
    for c in code:
        assert c.isalpha(), "Non-letter in airport code: " + repr(code)
        assert c.isupper(), "Lowercase letter in airport code: " + repr(code)
    return True
```

> - assert len(code) == 3  断言机场代码长度为3个字母
> - assert c.isalpha()    遍历机场代码的每个字符，断言字符是字母
> - assert c.isupper()   断言字符是大写字母

现在可以使用`code_repOK`函数来检查列表中的元素

```c
def airport_codes_repOK():
    for code in airport_codes:
        assert code_repOK(code)
    return True
```

```c
with ExpectError():
    assert airport_codes_repOK()
```

如果往列表中添加一个不符合要求的元素，如下，上面的检测报错。

```c
airport_codes["YMML"] = "Melbourne"
```

```c++
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/2308942452.py", line 2, in <cell line: 1>
    assert airport_codes_repOK()
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/480627665.py", line 3, in airport_codes_repOK
    assert code_repOK(code)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/865020192.py", line 2, in code_repOK
    assert len(code) == 3, "Airport code must have three characters: " + repr(code)
AssertionError: Airport code must have three characters: 'YMML' (expected)
```

当然，与其直接操作列表，还不如使用一个特殊函数来添加元素；这样还可以检查代码是否有效：

```c
def add_new_airport(code: str, city: str) -> None:
    assert code_repOK(code)
    airport_codes[code] = city
```

> 在添加元素的过程中就进行了检查

例如添加一个不符合要求的元素
```c
with ExpectError():
    add_new_airport("London-Heathrow", "LHR")
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/1427835309.py", line 2, in <cell line: 1>
    add_new_airport("London-Heathrow", "LHR")
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/2655039924.py", line 2, in add_new_airport
    assert code_repOK(code)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/865020192.py", line 2, in code_repOK
    assert len(code) == 3, "Airport code must have three characters: " + repr(code)
AssertionError: Airport code must have three characters: 'London-Heathrow' (expected)
```

不过，为了最大限度地进行检查，`add_new_airport()` 函数还将确保机场代码列表 在更改前后的正确表示。

```c
def add_new_airport_2(code: str, city: str) -> None:
    assert code_repOK(code)
    assert airport_codes_repOK()
    airport_codes[code] = city
    assert airport_codes_repOK()
```

测试：

```c
with ExpectError():
    add_new_airport_2("IST", "Istanbul Yeni Havalimanı")
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/4151824846.py", line 2, in <cell line: 1>
    add_new_airport_2("IST", "Istanbul Yeni Havalimanı")
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/2099116665.py", line 3, in add_new_airport_2
    assert airport_codes_repOK()
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/480627665.py", line 3, in airport_codes_repOK
    assert code_repOK(code)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/865020192.py", line 2, in code_repOK
    assert len(code) == 3, "Airport code must have three characters: " + repr(code)
AssertionError: Airport code must have three characters: 'YMML' (expected)
```

在程序中存在越多的断言，你能捕捉到错误的机会就越多-即使是针对特对领域的特定错误。此外，这些断言还记录了您在编程过程中所做的假设，因此有助于其他程序员理解您的代码并防止错误。

作为最后一个例子，让我们考虑一个相当复杂的数据结构----[红黑树](https://en.wikipedia.org/wiki/Red-black_tree)，一种自平衡的二叉搜索树。实现红黑树并不太难，但即使对于有经验的程序员来说，确保其正确性可能需要花费数小时的时间。然而，`repOK()` 方法不仅记录了所有的假设，还对这些假设进行了检查：

```c
class RedBlackTree:
    def repOK(self):
        assert self.rootHasNoParent()
        assert self.rootIsBlack()
        assert self.rootNodesHaveOnlyBlackChildren()
        assert self.treeIsAcyclic()
        assert self.parentsAreConsistent()
        return True

    def rootIsBlack(self):
        if self.parent is None:
            assert self.color == BLACK
        return True

    def add_element(self, elem):
        assert self.repOK()
        assert self.repOK()

    def delete_element(self, elem):
        assert self.repOK()
        assert self.repOK()
```

在这里，`repOK()` 是在 `RedBlackTree` 类的对象上运行的方法。它运行了五个不同的检查，每个检查都有它们自己的断言。每当添加或删除元素时，所有一致性检查都会自动运行。如果在这些检查中有任何错误，检查器都会找出它们 - 当然，前提是您通过足够多的模糊输入来运行这棵树。



## Static Code Checkers

在代码中使用静态类型检查程序也能获得`repOK()`断言的许多好处。例如，在Python中，只要正确声明参数类型，[MyPy](http://mypy-lang.org/)静态检查程序就能发现类型错误：

```c
typed_airport_codes: Dict[str, str] = {
    "YVR": "Vancouver",  # etc
}
```

如果现在添加一个键值对(没有值)，如下：

```c
typed_airport_codes[1] = "First"
```

MyPy会立即捕捉到错误：

```c
$ mypy airports.py
airports.py: error: Invalid index type "int" for "Dict[str, str]"; expected type "str"
```

静态地检查更高级的属性，比如机场代码由恰好三个大写字符组成，或者树是无环的，然而很快就会达到静态检查的限制。您仍然需要使用 `repOK()` 断言 ---最好与一个良好的测试用例生成器结合使用。



# A Fuzzing Architerctire

由于我们希望在接下来的章节中重用本章的一些部分，因此让我们以一种更易于重用、特别是更易于扩展的方式定义事物。为此，我们引入了一些类，以可重用的方式封装了上述功能。

## Runner Classes

首先引入的概念是”`Runner`“： 一个 执行给定输入的对象的 对象

> 有点绕口，Runner本身是一个`函数`，执行`给定输入的函数`。当然，不一定是函数，原文中用的是object(对象)

让我们用最基本的`Runner`开始，一个Runner通常提供一个 `run(input)` 方法，用于将输入传递给runner。`run()`方法返回一个二元组(result,outcome),在这里，result是一个值，代表指定runner的运行细节。outcome是是一个将结果分为三个类别的值：

- `Runner.PASS` - 测试通过。运行产生了正确的结果。
- `Runner.FAIL` - 测试失败。运行产生了不正确的结果。
- `Runner.UNRESOLVED` - 测试既没有通过也没有失败。这发生在运行无法进行的情况下 - 例如，输入无效。

```c
class Runner:
    """Base class for testing inputs."""

    # Test outcomes
    PASS = "PASS"
    FAIL = "FAIL"
    UNRESOLVED = "UNRESOLVED"

    def __init__(self) -> None:
        """Initialize"""
        pass

    def run(self, inp: str) -> Any:
        """Run the runner with the given input"""
        return (inp, Runner.UNRESOLVED)
```

作为一个基类(类比于Java中的)，Runner仅为在其上构建的更复杂的运行程序提供接口。 更具体地说，我们引入*子类*，*继承其超类的方法，以添加其他方法或覆盖继承的方法。*

下面是一个子类的示例：`PrintRunner`会简单的打印出给它的所有内容，从而重写继承的`run()`方法。

```c
class PrintRunner(Runner):
    def run(self, inp):
        """Print the given input"""
        print(inp)
        return (inp, Runner.UNRESOLVED)
```

```c
p = PrintRunner()
(result, outcome) = p.run("Some input")
```

结果就是传递的字符串。

![image](https://github.com/kksaert/fuzzbook/assets/83335903/c42a939f-285a-4e1d-bb0f-766acf0539d5)


`ProgramRunner`类将输入发送到程序的标准输入。 创建`ProgramRunner`对象时指定程序。

```c
class ProgramRunner(Runner):
    def __init__(self, program):
        """Initialize.  `program` is a program spec as passed to `subprocess.run()`"""
        self.program = program

    def run_process(self, inp=""):
        """Run the program with `inp` as input.  Return result of `subprocess.run()`."""
        return subprocess.run(self.program,
                              input=inp,
                              stdout=subprocess.PIPE,
                              stderr=subprocess.PIPE,
                              universal_newlines=True)

    def run(self, inp=""):
        """Run the program with `inp` as input.  Return test outcome based on result of `subprocess.run()`."""
        result = self.run_process(inp)

        if result.returncode == 0:
            outcome = self.PASS
        elif result.returncode < 0:
            outcome = self.FAIL
        else:
            outcome = self.UNRESOLVED

        return (result, outcome)
```

> - def __init__(self, program):构造函数，接受一个参数program，预期要传给下面的run_process函数
> - **`run_process` 方法：**负责使用 `subprocess.run()` 函数运行指定的程序，同时可以将输入 (`inp`) 传递给程序。函数使用了 `subprocess.PIPE` 来捕获标准输出和标准错误流，`universal_newlines=True` 参数则用于将输出文本解码为字符串。
> - **`run` 方法：**运行程序，并根据子进程的返回码确定程序执行的结果。它首先调用 `run_process` 方法来运行程序，然后根据返回码判断执行的结果。如果返回码为 0，表示程序成功执行，结果被设置为 "PASS"；如果返回码小于 0，表示程序执行出错，结果被设置为 "FAIL"；否则结果被设置为 "UNRESOLVED"。
>
> 这个类的目的是为了方便地运行其他程序，并根据子进程的返回码来判断程序执行的结果是成功、失败还是未确定。

下面是接收二进制输入的变

```c
class BinaryProgramRunner(ProgramRunner):
    def run_process(self, inp=""):
        """Run the program with `inp` as input.  Return result of `subprocess.run()`."""
        return subprocess.run(self.program,
                              input=inp.encode(), #此处对输入进行二进制转换
                              stdout=subprocess.PIPE,
                              stderr=subprocess.PIPE)

```

```c
cat = ProgramRunner(program="cat")
cat.run("hello")
```

让我们使用`cat`程序演示`ProgramRunner`，该程序将其输入复制到其输出。 我们看到`cat`的标准调用仅完成了工作，`cat`的输出与其输入相同：

![image](https://github.com/kksaert/fuzzbook/assets/83335903/8ee39fd1-3f22-47ef-8d02-c59ffa5d86f7)


## Fuzzer Classes

现在来实际定义一个产生数据的fuzzer，`Fuzzer基类`提供了一个核心方法`fuzz()`,上面的run()方法将`fuzz()`方法生成的数据提供给一个runner，并返回测试结果。`run()`方法在给定次数内重复执行过程

```c
class Fuzzer:
    """Base class for fuzzers."""

    def __init__(self) : #初始化方法
        pass  

    def fuzz(self) -> str: #生成fuzz数据
        return ""   

    def run(self, runner: Runner = Runner()) \
            -> Tuple[subprocess.CompletedProcess, Outcome]: 
        return runner.run(self.fuzz())

    def runs(self, runner: Runner = PrintRunner(), trials: int = 10) \
            -> List[Tuple[subprocess.CompletedProcess, Outcome]]:
        return [self.run(runner) for i in range(trials)]
```

> `def run(...)`:运行一个指定的runner，并将fuzz()函数产生的数据传入给runner.run()。它返回一个元组，包子进程的完成状态("subprocess.CompletedProcess")和fuzz结果(”outcome“)
>
> `def runs(...)`:和上面的run()相比，加了循环可以多次执行，且返回一个包含多个元组的列表(理解为二维数组)
>
> 

子类`RandomFuzzer`实现上述`fuzzer`函数的功能，

```c
class RandomFuzzer(Fuzzer):
    def __init__(self, min_length=10, max_length=100,
                 char_start=32, char_range=32):
        self.min_length = min_length
        self.max_length = max_length
        self.char_start = char_start
        self.char_range = char_range

    def fuzz(self):
        string_length = random.randrange(self.min_length, self.max_length + 1)
        out = ""
        for i in range(0, string_length):
            out += chr(random.randrange(self.char_start,
                                        self.char_start + self.char_range))
        return out
```

是面Fuzzer的子类，实现了随机生成字符串。

> **构造函数 (`__init__`)：**：接受四个参数，依次是最小长度，最大长度，起始字符值，字符范围
>
> `fuzz()`:先生成一个随机的字符串长度，其范围在 `min_length` 到 `max_length` 之间其范围在 `min_length` 到 `max_length` 之间

```c
random_fuzzer = RandomFuzzer(min_length=20, max_length=20)
for i in range(10):
    print(random_fuzzer.fuzz())
```

```c
*5!&138$==2%*$3* -59
54.-!<18<98*=9?/(8:!
",5>5:>!&1'=5%9//>(.
/&5(0++60$.($0<,1,5,
*:>8+3,:"-+686?>?+23
#"%3-557":!%9-<=3<!"
&86% 1&>+.8)>4/>'9;?
$=&8!18+ 78636+6>1"?
0*.=&7?25,-=63#266.*
###.2,832>/9!%9%*'&+
```

现在，可以将生成的输入发送到之前的`cat`runner，来对cat程序进行fuzz

```c
for i in range(10):
    inp = random_fuzzer.fuzz()
    result, outcome = cat.run(inp)
    assert result.stdout == inp
    assert outcome == Runner.PASS
```

上面这是一种简单的写法，别忘了Fuzz基类的run()方法和runs()方法。

```c
random_fuzzer.run(cat)
```

```c
(CompletedProcess(args='cat', returncode=0, stdout='?:+= % <1<6$:(>=:9)5', stderr=''),
 'PASS')
```

也可以重复执行：

```c
random_fuzzer.runs(cat, 10)
```

```c
[(CompletedProcess(args='cat', returncode=0, stdout='3976%%&+%6=(1)3&3:<9', stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout='33$#42$ 11=*%$20=<.-', stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout='"?<\'#8 </:*%9.--\'97!', stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout="/0-#(03/!#60'+6>&&72", stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout="=,+:,6'5:950+><3(*()", stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout=" 379+0?'%3137=2:4605", stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout="02>!$</'*81.#</22>+:", stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout="=-<'3-#88*%&*9< +1&&", stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout='2;;0=3&6=8&30&<-;?*;', stderr=''),
  'PASS'),
 (CompletedProcess(args='cat', returncode=0, stdout='/#05=*3($>::#7!0=12+', stderr=''),
  'PASS')]
```



## Lessons Learned

- 随机生成输入（“fuzzing”）是一种简单，经济高效的方法，可以快速测试任意程序的健壮性。
- fuzzer发现的错误主要是由于程序处理输入中的错误和不足。
- 要捕获错误，请尽可能多地使用*一致性检查器*。

## Next Steps

接下来可以阅读

- [use *mutations* on existing inputs to get more valid inputs]()
- [use *grammars* to specify the input format and thus get many more valid inputs]()
- [reduce *failing inputs* for efficient debugging]()

## Background

有关生成软件测试的书籍很少（这就是我们编写这本书的原因）。 尽管有一些关于*模糊测试的书籍*，但也基于本章介绍的基本模糊测试技术：

- "Fuzzing – Brute Force Vulnerability Discovery"：一书涵盖了广泛的模糊测试领域，包括文件，网页，环境变量和网络协议。 作者从Microsoft的模糊测试中获得了很多经验，并包括了许多用于Windows和UNIX程序的现成工具。 工具已经有些老化，但是原理仍然存在。

- "Fuzzing for Software Security Testing and Quality Assurance" [Takanen et al, 2008](http://us.artechhouse.com/Fuzzing-for-Software-Security-Testing-and-Quality-Assurance-Second-Edition-P1930.aspx),现在出版，2018年第二版，涵盖了广泛的模糊测试工具和检测技术 ; 其作者从安全测试和漏洞发现中获得了丰富的经验。 这可以说是该领域最全面，最新的书籍之一。

特别是在[An Empirical Study of the Reliability of UNIX Utilities](https://doi.org/10.1145/96267.96279)这篇paper中，介绍了有关fuzz的开创性工作(介绍术语和方法)，作为该领域的基础，这对于任何对模糊测试和鲁棒性测试感兴趣的人都必须阅读，其观察与30年前一样有效。



## Exercise

米勒等人发现的错误之一，[Miller *et al*, 1990](https://doi.org/10.1145/96267.96279)，涉及 *troff* 排版系统，*Troff* 接受由行组成的文本作为输入； 以点开头的行（`.`）包含排版命令，如

```c
.NH
Some Heading
.LP
Some paragraph
```

这将产生文本(使用`nroff -ms`)

Miller等人当时发现， 如果输入是以下情况，*troff* 会运行失败

1. 输入序列`\D`（反斜杠+ D），后跟不可打印字符
2. ASCII范围为128–255的字符（即设置了第8位），后跟换行符
3. 单个点（`.`）后跟换行符。

## Exercise 1: Simulate(模拟) Troff

对于上述每一，编写一个Python函数`f(s)`，如果`s`满足失败条件，该函数将失败。

第一点实现：不可打印字符有很多，另起一行，退格等。

```c
import string
def no_backslash_d(inp):
    pattern = "\\D"
    index = inp.find(pattern)  #找到inp字符串中第一次出现pattern的索引位置，并将结果保存在index中。
    if index < 0 or index + len(pattern) >= len(inp):
        return True
    c = inp[index + len(pattern)]
    assert c in string.printable # string.printable包含了所有可打印字符的字符串，如果 c 不在其中，断言会触发异常。
```

> `index < 0 or index + len(pattern) >= len(inp)` ：如果没有找到（`index`小于0），满足要求，`index + len(pattern)` 计算的是patter结束的索引位置。如果大于len(inp)的长度，意味着patter没有足够的空间在inp中出现，符合要求。

```c
with ExpectError():
    no_backslash_d("\\D\0")
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/1873852117.py", line 2, in <cell line: 1>
    no_backslash_d("\\D\0")
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/2739137185.py", line 7, in no_backslash_d
    assert c in string.printable
AssertionError (expected)
```

测试成功。

第二个条件：

```c
def no_8bit(inp):
    for i in range(len(inp) - 1):
        assert ord(inp[i]) <= 127 or inp[i + 1] != '\n'
    return True
```

> 循环遍历输入的字符串，`ord(inp[i]) <= 127`：检查当前字符的 ASCII 值是否不超过 127。`inp[i + 1] != '\n'`：检查当前字符后面的字符是否为换行符 `\n`

测试：

```
with ExpectError():
    no_8bit("ä\n")
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/4044610929.py", line 2, in <cell line: 1>
    no_8bit("ä\n")
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_1242/986149540.py", line 3, in no_8bit
    assert ord(inp[i]) <= 127 or inp[i + 1] != '\n'
AssertionError (expected)
```

第三个条件：

```c
def no_dot(inp):
    assert inp != ".\n"
    return True
```



## Exercise 2: Run Simulated Troff

创建一个类`TroffRunner`作为`Runner`的子类，使用`Fuzzer`运行它。 确保`Fuzzer`对象产生整个字符范围。 计算失败的频率。

```c
class TroffRunner(Runner):
    def __init__(self): #用于跟踪不同断言失败的次数
        self.no_backslash_d_failures = 0
        self.no_8bit_failures = 0
        self.no_dot_failures = 0

    def run(self, inp):
        try:
            no_backslash_d(inp)
        except AssertionError:
            self.no_backslash_d_failures += 1

        try:
            no_8bit(inp)
        except AssertionError:
            self.no_8bit_failures += 1

        try:
            no_dot(inp)
        except:
            self.no_dot_failures += 1

        return inp
```

```c
random_fuzzer = RandomFuzzer(char_start=0, char_range=256, max_length=10)
troff_runner = TroffRunner()
```

```c
trials = 100000
for i in range(trials):
    random_fuzzer.run(troff_runner)
```

```c
troff_runner.no_backslash_d_failures

5 
```

```c
troff_runner.no_8bit_failures

1717
```

```c
troff_runner.no_dot_failures

0
```



## Exercise 3: Run Real Troff

使用`BinaryProgramRunner`，应用在实际`troff`程序上配置的模糊器。 检查是否可以产生任何输出代码非零的运行，表示失败或崩溃。

```c
real_troff_runner = BinaryProgramRunner("troff")
for i in range(100):
    result, outcome = random_fuzzer.run(real_troff_runner)
    if outcome == Runner.FAIL:
        print(result)
```

很不幸，在这个阶段很不可能在troff中找到漏洞。就像大多数其他开源代码一样，它已经通过这种方式进行了模糊测试，所有已发现的错误都已经修复。
